### PR TITLE
BE-044 Search for discounts (Elizaveta Dubogryz)

### PR DESCRIPTION
**Pull request by el-dub**

### CODE REVIEW

**Review by rudik-byte — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java [hunk 1]
```diff
@@ -112,6 +113,17 @@ public DiscountDTO restoreById(UUID id) {
         return discountMapper.getDTO(discount);
     }
 
+    @Override
+    public List<DiscountDTO> search(Integer size, String searchText) {
+        log.debug("Getting list of all Discounts by searchText");
+        List<Discount> discounts = discountRepository
+                .findAll(prepareSearchPredicate(searchText), Sort.by("viewNumber"));
+        Page<Discount> discountPage = PageableExecutionUtils
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: Paging should happen on the DB call, not after

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java [hunk 2]
```diff
@@ -168,4 +180,20 @@ private Predicate preparePredicateForFindingAll(DiscountFilter filter) {
                         .append(filter.getVendorIds(), QDiscount.discount.vendor.id::in)
                         .buildAnd());
     }
+
+    private Predicate prepareSearchPredicate(String searchText) {
+        List<Predicate> searchPredicates = Pattern
+                .compile(" ")
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: replace with string.split or use StringUtils for that

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java [hunk 3]
```diff
@@ -168,4 +180,20 @@ private Predicate preparePredicateForFindingAll(DiscountFilter filter) {
...
+                .splitAsStream(searchText)
+                .filter(word -> word.length() >= 3)
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: move to constants and name carefully for clear understanding what is that a magic number

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java (hunk 1 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: +

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java (hunk 2 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: +

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java (hunk 3 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: +

**Review by rudik-byte — DISMISSED** [review]

**Review by rudik-byte — DISMISSED** [review]

**Review by Hanna-Kalashnyk — DISMISSED** [review]

File: src/main/java/com/exadel/discount/repository/DiscountCustomRepository.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.querydsl.core.types.Predicate;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.UUID;
+
+@Repository
+public interface DiscountCustomRepository {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/repository/DiscountCustomRepository.java]:
```suggestion
public interface QueryFactoryDiscountRepository {
```

File: src/main/java/com/exadel/discount/repository/impl/DiscountCustomRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import com.exadel.discount.repository.DiscountCustomRepository;
+import com.querydsl.core.types.Predicate;
+import com.querydsl.jpa.impl.JPAQueryFactory;
+import org.springframework.data.domain.Pageable;
+
+import javax.persistence.EntityManager;
+import java.util.List;
+import java.util.UUID;
+
+public class DiscountCustomRepositoryImpl implements DiscountCustomRepository {
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/repository/impl/DiscountCustomRepositoryImpl.java]: I don't see annotations. Who and how creates this class and pass em to constructor?

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java [hunk 6]
```diff
@@ -166,4 +177,19 @@ private Predicate preparePredicateForFindingAll(DiscountFilter filter) {
                         .append(filter.getVendorIds(), QDiscount.discount.vendor.id::in)
                         .buildAnd());
     }
+
+    private Predicate prepareSearchPredicate(String searchText) {
+        List<Predicate> searchPredicates = Stream
+                .of(StringUtils.split(searchText, " "))
```
Comment (SergeyKoval) [inline src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]:
```suggestion
                .of(StringUtils.split(searchText, StringUtils.SPACE))
```

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/discount/repository/DiscountCustomRepository.java (hunk 4 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/repository/DiscountCustomRepository.java]: +

File: src/main/java/com/exadel/discount/repository/impl/DiscountCustomRepositoryImpl.java (hunk 5 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/repository/impl/DiscountCustomRepositoryImpl.java]: +

File: src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java (hunk 6 above)
Comment (el-dub) [reply src/main/java/com/exadel/discount/service/impl/DiscountServiceImpl.java]: +

**Review by rudik-byte — DISMISSED** [review]

**Review by SergeyKoval — DISMISSED** [review]

**Review by rudik-byte — DISMISSED** [review]

**Review by rudik-byte — APPROVED** [review]

**Review by SergeyKoval — APPROVED** [review]

**Pull request merged**